rasa_sdk/knowledge_base/actions.py

```diff
@@ -0,0 +1,372 @@
+
+        for attr in self.knowledge_base.schema[object_type][SCHEMA_KEYS_ATTRIBUTES]:
+            attr_val = tracker.get_slot(attr)
+            if attr_val is not None:
+                attributes.append({"name": attr, "value": attr_val})
+
+        return attributes
+
+    def _reset_attribute_slots(self, tracker, object_type):
+        """
+        Reset all attribute slots of the current object type.
+
+        If the user is saying something like "Show me all restaurants with Italian
+        cuisine.", the NER should detect "restaurant" as "object_type" and "Italian" as
+        "cuisine" object. So, we should filter the restaurant objects in the
+        knowledge base by their cuisine (= Italian). When listing objects, we check
+        what attributes are detected by the NER. We take all attributes that are set,
+        e.g. cuisine = Italian. If we don't reset the attribute slots after the request
+        is done and the next utterance of the user would be, for example, "List all
+        restaurants that have wifi.", we would have two attribute slots set: "wifi" and
+        "cuisine". Thus, we would filter all restaurants for two attributes now:
+        wifi = True and cuisine = Italian. However, the user did not specify any
+        cuisine in his request. To avoid that we reset the attribute slots once the
+        request is done.
+
+        Args:
+            object_type: the object type
+            tracker: the tracker
+
+        Returns: list of slots
+        """
+        slots = []
+
+        if object_type not in self.knowledge_base.schema:
+            return slots
+
+        for attr in self.knowledge_base.schema[object_type][SCHEMA_KEYS_ATTRIBUTES]:
+            attr_val = tracker.get_slot(attr)
+            if attr_val is not None:
+                slots.append(SlotSet(attr, None))
+
+        return slots
+
+
+class ActionQueryKnowledgeBase(ActionKnowledgeBase):
+    """
+    Action that queries the knowledge base for objects and attributes of an object.
+    The action needs to be inherited and the knowledge base needs to be set.
+    In order to actually query the knowledge base you need to:
+    - create your knowledge base
+    - add mandatory slots to the domain file: 'object_type', 'attribute', 'mention'
+    - create NLU data where the required objects are annotated
+    - create a story that includes this action
+    - add the intent and action to domain file
+    """
+
+    def __init__(self, knowldege_base):
+        super(ActionQueryKnowledgeBase, self).__init__(knowldege_base)
+
+    def name(self):
+        return "action_query_knowledge_base"
+
+    def utter_rephrase(self, dispatcher):
+        """
+        Utters a response to the user that indicates that something went wrong. It
+        asks the user to rephrase his request.
+
+        Args:
+            dispatcher: the dispatcher
+        """
+        dispatcher.utter_message("Sorry, I did not get that. Can you please rephrase?")
```

Comment:
Let's not hardcode the messages, but rather ask a user to create utter templates with certain names